libsolutil/LP.cpp

```diff
@@ -0,0 +1,1613 @@
+{
+	if (_expr.arguments.empty())
+		m_state.back().clauses.emplace_back(Clause{vector<Literal>{*parseLiteral(_expr)}});
+	else if (_expr.name == "=")
+	{
+		// Try to see if both sides are linear.
+		optional<vector<rational>> left = parseLinearSum(_expr.arguments.at(0));
+		optional<vector<rational>> right = parseLinearSum(_expr.arguments.at(1));
+		if (left && right)
+		{
+			vector<rational> data = *left - *right;
+			data[0] *= -1;
+			Constraint c{move(data), _expr.name == "="};
+			if (!tryAddDirectBounds(c))
+			{
+				size_t index = addConstraint(move(c));
+				m_state.back().clauses.emplace_back(Clause{vector<Literal>{Literal{Literal::Constraint, index}}});
+			}
+		}
+		else if (_expr.arguments.at(0).arguments.empty() && isBooleanVariable(_expr.arguments.at(0).name))
+			addBooleanEquality(*parseLiteral(_expr.arguments.at(0)), _expr.arguments.at(1));
+		else if (_expr.arguments.at(1).arguments.empty() && isBooleanVariable(_expr.arguments.at(1).name))
+			addBooleanEquality(*parseLiteral(_expr.arguments.at(1)), _expr.arguments.at(0));
+		else
+		{
+			Literal const& newBoolean = *parseLiteral(declareInternalBoolean());
+			addBooleanEquality(newBoolean, _expr.arguments.at(0));
+			addBooleanEquality(newBoolean, _expr.arguments.at(1));
+		}
+	}
+	else if (_expr.name == "and")
+	{
+		addAssertion(_expr.arguments.at(0));
+		addAssertion(_expr.arguments.at(1));
+	}
+	else if (_expr.name == "or")
+	{
+		// We could try to parse a full clause here.
+		Literal left = parseLiteralOrReturnEqualBoolean(_expr.arguments.at(0));
+		Literal right = parseLiteralOrReturnEqualBoolean(_expr.arguments.at(1));
+		if (left.kind == Literal::Constraint && right.kind == Literal::Constraint)
+		{
+			// We cannot have more than one constraint per clause.
+			right = *parseLiteral(declareInternalBoolean());
+			addBooleanEquality(right, _expr.arguments.at(1));
+		}
+		m_state.back().clauses.emplace_back(Clause{vector<Literal>{left, right}});
+	}
+	else if (_expr.name == "not")
+	{
+		Literal l = negate(parseLiteralOrReturnEqualBoolean(_expr.arguments.at(0)));
+		m_state.back().clauses.emplace_back(Clause{vector<Literal>{l}});
+	}
+	else if (_expr.name == "<=")
+	{
+		optional<vector<rational>> left = parseLinearSum(_expr.arguments.at(0));
+		optional<vector<rational>> right = parseLinearSum(_expr.arguments.at(1));
+		if (!left || !right)
+			return;
+
+		vector<rational> data = *left - *right;
+		data[0] *= -1;
+		Constraint c{move(data), _expr.name == "="};
+		if (!tryAddDirectBounds(c))
+		{
+			size_t index = addConstraint(move(c));
+			m_state.back().clauses.emplace_back(Clause{vector<Literal>{Literal{Literal::Constraint, index}}});
+		}
+	}
+	else if (_expr.name == ">=")
+		addAssertion(_expr.arguments.at(1) <= _expr.arguments.at(0));
+	else if (_expr.name == "<")
+		addAssertion(_expr.arguments.at(0) <= _expr.arguments.at(1) - 1);
+	else if (_expr.name == ">")
+		addAssertion(_expr.arguments.at(1) < _expr.arguments.at(0));
+}
```

Comment:
Should there be here a failure case? This way we just ignore unparsed assertions.

Reply:
Yes, this is still a todo, but in general, since the whole solver can only detect infeasibility (it approximates integers by rationals), ignoring assertions that it does not support only reduces precision but is not a problem for correctness.